### PR TITLE
93 allow separate data weighting by partition

### PR DESCRIPTION
**Pull request by Rick-Methot-NOAA**
## Concisely (20 words or less) describe the issue
This change allows for the D-M parameters to be partition specific.
## Please Link issue(s)
#93 
resolves #[add issue number number], resolves #[optional second issue number if more than 1 issue addressed.]
incremental progress here
## What tests have been done? Upload any model input files created for testing in a zip file, if possible.
test input files attached.  Two dummy observations added that use different D-M parameter
## Has any new code been documented?
yes
If not, please add documentation before submitting the Pull Request.
- [X] I have documented any new code added (or no new code was added)

## is there an input change for users to Stock Synthesis? 
conditional input change invoked by the use_lengthcomp flag
- [x] Yes, there was an input change
conditional input change invoked by the use_lengthcomp flag

If so, please provide an example of the new inputs needed.

```
2 # use length composition data (0/1/2) where 2 invokes new comp_comtrol format
#_fleet partition mintailcomp addtocomp combM+F CompressBins CompError ParmSelect minsamplesize
 0 0 0.0001 4e-07 0 0 0 0 1
 1 1 0.0001 2e-07 0 0 1 1 1
 1 2 0.0001 3e-07 0 0 1 2 1
 -9999 1 0.0001 1e-07 0 0 0 0 1
```

## Check which is true. This PR requires:

- [ ] no further changes to r4ss
- [ ] no further changes to the manual
- [ ] no further changes to SSI (the SS3 GUI)
- [ ] no further changes to the stock synthesis change log (new features, bug reports)

## Describe any changes in r4ss/SS3 manual/SSI that are needed (if not checked):
r4ss modification needed for the fit_lencomp table to add three columns for the comp logL method and the potential D-M and MVT parameters;
[data_weighting.zip](https://github.com/nmfs-stock-synthesis/stock-synthesis/files/10249520/data_weighting.zip)

code needs to be mimicked in the age comp and Sz comp sections
## If changes are needed in the change log, please fill in the table here:
see the Project Board

### CODE REVIEW

**Comment by Rick-Methot-NOAA** [comment]
I ran hake_2018 locally to try to detect the reason for the GHA failure.  Some tiny numerical differences and change in the parameter name for the D-M parameters were all that I noticed.

**Comment by Rick-Methot-NOAA** [comment]
Before I add this partition capability to the age composition and gen_size composition, let me ask:  is it needed?  How likely is it that a user would have age comp for both discarded and retained fish?  For lengths, note that a user already could do discard lengths as length comp and retained lengths as gen_size, each with their own D-M parameters.
Rick

**Comment by iantaylor-NOAA** [comment]
@Rick-Methot-NOAA, thanks for working on this.
I put it on the agenda for the NWFSC/SWFSC assessment meeting tomorrow morning to get feedback on the need to expand beyond length comps--I'm guessing that the answer is no.
I'll try to review the changes tomorrow or Thursday depending on travel timing and snow.

**Comment by iantaylor-NOAA** [comment]
Feedback at assessment team meeting was that there are some assessments with a few years with ages from discarded fish (nobody could remember which stocks), but that it shouldn't be a problem to apply the status-quo assumption of equal weight to those discards as to the associated retained ages given that the influence of those samples is presumably pretty low.

There is some interested (at least from me, but I think maybe others), in trying out the separate weighting for discard lengths using the beta version of SS3 to see how different the estimates are and provide feedback in advance of the release. I should be able to do that tomorrow along with reviewing the Pull Request.

**Review by iantaylor-NOAA — DISMISSED** [review]
Looks good overall. I had some trouble debugging a test model (working version is attached) because of the following two issues:
* I had the retained (part = 2) before discards (part = 1) in the expanded table of length comp controls which caused an error, but switching the order of those lines also didn't work because then the ParmSelect values were out of order (I should have caught that). So the documentation should make clear that the partitions need to be entered in numerical order within each fleet. I think it's already documented that the ParmSelect values need to be in increasing order.
* The echoinput reports the placeholder values for all fleet/partition combinations that aren't represented in the table with a warning triggered by each one. In the case of the attached model (modified from BigSkate_2019 in test-models), I have data for discards and retained in fleet 1, but not for part = 0, so the row in echoinput.sso is as follows:
  ```fleet partition  mintail  addtocomp conbgender accumbin  comperr  comperrparm
  1 0 0 0 0 0 0 0 0  #_fleet: 1 Fishery_current
  Warning 1 :  minimum sample size for length comps must be > 0; minimum sample size set to 
  0.001 
  ```
Perhaps a simple fix would be to provide placeholder values with min sample size > 0 so the warning didn't get triggered. You would still see those values echoed, since the model doesn't yet know what fleets/partitions have length comps.


[bigskate_partition_weighting.zip](https://github.com/nmfs-stock-synthesis/stock-synthesis/files/10297573/bigskate_partition_weighting.zip)

**Comment by iantaylor-NOAA** [comment]
One additional comment:
These lines 
https://github.com/nmfs-stock-synthesis/stock-synthesis/blob/9f5459c77993ae541cb017eff2190c449781cc0a/SS_write_report.tpl#L360-L361
used to be part of a simple table near the top of Report.sso because `Comp_Err_L` and `Comp_Err_L2` were vectors:
```
Fleet_name:  Fishery_current Discard_historical Fishery_historical Fishery_tribal WCGBTS Triennial
Fleet_type:  1 1 1 1 3 3
Fleet_area:  1 1 1 1 1 1
Lencomp_error_type:  0 0 0 0 0 0
Lencomp_error_parms:  0 0 0 0 0 0
Agecomp_error_type:  0 0 0 0 0 0
Agecomp_error_parms:  0 0 0 0 0 0
```

However, those variables (`Comp_Err_L` and `Comp_Err_L2`) are now matrices with a row for each partition, so the table looks like this, where it's not clear to the user what each row means:
```
Fleet_name:  Fishery_current Discard_historical Fishery_historical Fishery_tribal WCGBTS Triennial
Fleet_type:  1 1 1 1 3 3
Fleet_area:  1 1 1 1 1 1
Lencomp_error_type:  0 0 0 0 1 1
 1 0 0 0 0 0
 1 0 0 0 0 0
Lencomp_error_parms:  0 0 0 0 3 4
 1 0 0 0 0 0
 2 0 0 0 0 0
Agecomp_error_type:  1 0 0 0 1 0
Agecomp_error_parms:  5 0 0 0 6 0
```

This matrix output is present even for models that don't use either the D-M likelihood or separate weighting by partition. 
Could we consider a different format for this table to make it clear what each row represents?

**Comment by Rick-Methot-NOAA** [comment]
Ian,
Thanks for the very thorough review.  I'll be able to get all of this updated this week.

File: SS_readdata_330.tpl [hunk 1]
```diff
@@ -1339,50 +1340,127 @@
   Comp_Err_parmloc.initialize();
   min_sample_size_L.initialize();
   DM_parmlist.initialize();
-
+  comp_control_L_count = -1;
   if (use_length_data > 0)
   {
     echoinput << "#_now read for each fleet info for processing the length comps:" << endl;
     echoinput << "#_mintailcomp: upper and lower distribution for females and males separately are accumulated until exceeding this level." << endl;
```
[the remark this answers is not in the document]
Comment (Rick-Methot-NOAA) [reply SS_readdata_330.tpl:1347]: I made it more obvious that user should fill all higher numbered fleets by using a negative fleet input.

**Comment by iantaylor-NOAA** [comment]
I finally got r4ss working with partition weighting for one model in https://github.com/r4ss/r4ss/commit/bdf9ed671b1701458973d7b7704292d82e8e9d03. The changes were extensive because there was layer upon layer of clunky stuff related to D-M likelihoods and parameters that was overdue to get cleaned up (we were still reading the data file to get info on which D-M parameter went with which fleet because many SS3 versions ago that info wasn't available in the Report.sso).

If the r4ss changes pass the r4ss tests and I don't discover any big issues when testing on a few older versions of SS3, then I will merge the changes into the r4ss main branch so that we can get the SS3 checks working.

Alternatively we could create a branch of the "workflows" repo and change `test-r4ss-with-ss3` to point to the "partition-weighting" branch in r4ss AND change the `call-test-r4ss-with-ss3` workflow in this "stock-synthesis" branch (`93-allow-separate-data-weighting-by-partition`) to use the branch of "workflows". But all that seems like more work than just making sure I didn't break r4ss and merging that first.

**Comment by iantaylor-NOAA** [comment]
The r4ss fix for these partition weighting changes is finally complete in a new branch and @e-gugliotti-NOAA is on deck to review the Pull Request https://github.com/r4ss/r4ss/pull/766 and merge tomorrow which will facilitate the final testing via SS3 github actions.

I also just pushed an addition commit 48af932 to this SS3 branch which provides a default minsample size value 0.001 to avoid the warning `minimum sample size for length comps must be > 0; minimum sample size set to 0.001` appearing for combinations of fleet and partition which don't have minsamplesize provided because they don't have any data. @Rick-Methot-NOAA, let me know if that changes doesn't look OK to you--I didn't bother to create an additional branch and pull request to this branch.

**Comment by Rick-Methot-NOAA** [comment]
Good addition regarding minsampsize.  User could accomplish this by filling all rows with the -fleet option, but they may not do that, so your addition is safeguard.

**Comment by e-gugliotti-NOAA** [comment]
@iantaylor-NOAA I just re-ran the test r4ss workflow and it is not passing. Looks like a dplyr issue is at least one problem. [See failed workflow here](https://github.com/nmfs-stock-synthesis/stock-synthesis/actions/runs/4049469021/jobs/6975660068)

**Review by iantaylor-NOAA — APPROVED** [review]
approving changes since last review to unlock merge

**Comment by iantaylor-NOAA** [comment]
@Rick-Methot-NOAA, I think you can merge this now.

@e-gugliotti-NOAA got the remaining r4ss bugs worked out and the automated checks of the r4ss on the test-models repository models are now passing as well as local tests on a model with partition weighting.

We may want to add a model to the test-set that has partition weighting, or at least switch one of the existing models to make sure of the input associated with data file setting:
`2 # use length composition data (0/1/2) where 2 invokes new comp_control format`

I'll add this to the list in https://github.com/nmfs-stock-synthesis/test-models/issues/29.

**Comment by Rick-Methot-NOAA** [comment]
Thanks Ian and Elizabeth.  Seems like the r4ss work was bigger than the SS3 work.  I will merge now and I endorse adding a model with this feature to the test set.  But let's also think about other features that are not represented in the test set and implement them in the same test model.

**Pull request merged**